File: Others/02. Nth Stair Problem/frogRoutes.java
/* Special class of Program with O(exp(n)) naive solution but has O(n) soultion using 
    dynamic programming and O(log(n)) optimized solution using matrix multiplication.
   Required Property : T(n) = Σ±Ci*T(n-i), where Ci could be 0 and Range of i << n.
   Example : A frog can leap 2,3 or 5 step in one jump. Given a stairs of n steps, 
    determine the number of different possible routes the frog can take.

   Use as below:
    G:\Temp\Programming> javac frogRoutes.java
    G:\Temp\Programming> java frogRoutes
*/

class xtremeOptimisation {
    private long[] baseCase = null;
    private long[][] Matrix = null;

    /**
     * Computes T(n) with log(n) time complexity.
     * @param baseCase         := value of T(i) for n = 0 case
     * @param nthElementVector := value of T(n) in terms of T(n-i)s. So for
     *                          T(n) = T(n-2) + T(n-3), it would be { 0, 1, 1 }
     * @param n                := Value of n
     * @return long value of T(n)
     */
    public static long compute(int[] baseCase, int[] nthElementVector, int n) {
        if (baseCase.length > nthElementVector.length)
            return 0;

        xtremeOptimisation Ins = new xtremeOptimisation(baseCase, nthElementVector);
        long[][] resultMatrix = Ins.computeNthPowerMatrix(Ins.Matrix, n);
        long[] resultVector = Ins.multiplyVectorWithMatrix(Ins.baseCase, resultMatrix);

        if (resultVector == null || resultVector.length < 1)
            return 0;
        return resultVector[0];
    }

    private xtremeOptimisation(int[] baseCase, int[] nthElementVector) {
        this.baseCase = new long[nthElementVector.length];
        for (int i = 0; i < baseCase.length; i++)
            this.baseCase[i] = baseCase[i];

        this.Matrix = new long[nthElementVector.length][nthElementVector.length];
        for (int i = 0; i < Matrix.length; i++) {
            Matrix[i][0] = nthElementVector[i];
            if (i < Matrix.length - 1)
                Matrix[i][i + 1] = 1;
        }
    }

    private long[][] computeNthPowerMatrix(long[][] M, int n) {
        if (n < 0 || M==null || M.length == 0 || M.length != M[0].length)
            return null;

        long[][] resultMatrix = new long[M.length][M.length];
        for (int i = 0; i < resultMatrix.length; i++)
            resultMatrix[i][i] = 1;

        long[][] Mx = M.clone();
        for (int i = n; i > 0; i /= 2) {
            if (i % 2 == 1)
                resultMatrix = multiply2Matrices(resultMatrix, Mx);
            Mx = multiply2Matrices(Mx, Mx);
        }
        return resultMatrix;
    }

    private long[][] multiply2Matrices(long[][] A, long[][] B) {
        long[][] M = new long[A.length][B[0].length];
        for (int i = 0; i < A.length; i++) {
            M[i] = multiplyVectorWithMatrix(A[i], B);
        }
        return M;
    }

    private long[] multiplyVectorWithMatrix(long[] V, long[][] M) {
        long[] resVector = null;
        if (M == null || V.length != M.length)
            return resVector;
        else
            resVector = new long[M[0].length];

        for (int i = 0; i < M.length; i++) {
            for (int j = 0; j < M[i].length; j++) {
                resVector[j] += V[i] * M[i][j];
            }
        }
        return resVector;
    }
}

public class frogRoutes {
    public static void main(String[] args) {
        int[] baseCase = { 1, 0, 0, 0, 0 };
        int[] vec = { 0, 1, 1, 0, 1 };
        long result = xtremeOptimisation.compute(baseCase, vec, 100);

        System.out.println(result);
    }
}

File: Others/02. Nth Stair Problem/frogRoutes.ml
(* Special class of Program with O(exp(n)) naive solution but has O(n) soultion using 
    dynamic programming and O(log(n)) optimized solution using matrix multiplication.
   Required Property : T(n) = Σ±Ci*T(n-i), where Ci could be 0 and Range of i << n.
   Example : A frog can leap 2,3 or 5 step in one jump. Given a stairs of n steps, 
    determine the number of different possible routes the frog can take.
 
   Use as below:
    G:\Temp\Programming> ocaml
    # #use "frogRoutes.ml";;
   OR
    G:\Temp\Programming> ocaml frogRoutes.ml
*)

(*Gives Execution time of any function*)
let time f x =
  let t = Sys.time() in
  let fx = f x in
  Printf.printf "Execution time: %fs\n" (Sys.time() -. t);
  fx;;


(* O(exp(n)) naive solution.*)
  let rec frogSteps n = 
    if (n < 0) then 0 else
    if (n = 0) then 1 else
    (frogSteps (n-2)) + (frogSteps (n-3)) + (frogSteps (n-5));;

  let getFrogSteps n = Printf.printf "O(exp(n)) naive solution\n%d\n" (frogSteps n);;
  time getFrogSteps 50;;


(* O(n) dynamic programming solution.*)
  let frogSteps_n n =
    let rec frogStep_n (tn_1, tn_2, tn_3, tn_4, tn_5) i =
      if (i < 0) then 0 else
      if (i = 0) then tn_1 else
      frogStep_n (tn_2+tn_3+tn_5, tn_1, tn_2, tn_3, tn_4) (i-1)
    in frogStep_n (1,0,0,0,0) n;;

  let getFrogSteps_n n = Printf.printf "\nO(n) dynamic programming solution\n%d\n" (frogSteps_n n);;
  time getFrogSteps_n 50;;


(* O(log(n)) optimized solution using matrix multiplication.*)
  let frogSteps_logN n =
    let baseCase = [1;0;0;0;0] in
    let tn = [0;1;1;0;1] in
    let rec vector i j n =
      if (j >= n) then [] else
      if (j = i)
        then 1::vector i (j+1) n
        else 0::vector i (j+1) n in
    let idMatrix n =
      let rec matrix n i =
        if (i >= n) then [] else
        (vector i 0 n)::(matrix n (i+1)) in
      matrix n 0 in
    let inMatrix vec =
      let rec getmatrix n vec i = match vec with
        | [] -> []
        | h::rest -> (h::(vector (i+1) 1 n))::(getmatrix n rest (i+1)) in
      getmatrix (List.length vec) vec 0 in
    let rec mapn f lists = assert (lists <> []);
      if List.mem [] lists then []
      else f (List.map List.hd lists) :: mapn f (List.map List.tl lists) in
    let vector_multiply vec mat =
      mapn (fun column -> List.fold_left (+) 0 (List.map2 ( * ) vec column)) mat in
    let matrix_multiply m1 m2 =
      List.map (fun row -> mapn (fun column -> List.fold_left (+) 0 (List.map2 ( * ) row column)) m2) m1 in
    let rec matrixPow mtx n =
      if (n<=0) then idMatrix (List.length mtx) else
      if (n mod 2 = 0) then matrixPow (matrix_multiply mtx mtx) (n/2) else
      matrix_multiply mtx (matrixPow (matrix_multiply mtx mtx) (n/2)) in
    let finalVector = vector_multiply baseCase (matrixPow (inMatrix tn) n) in
    let result = match finalVector with [] -> 0 | head::rest -> head in
    result;;

  let getFrogSteps_logN n = Printf.printf "\nO(log(n)) optimized solution using matrix multiplication\n%d\n" (frogSteps_logN n);;
  time getFrogSteps_logN 50;;